test/gradcheck.jl

```diff
@@ -180,6 +180,11 @@ end
   @test back([nothing]) == ([nothing for _ in 1:3], nothing)
 end
 
+@testset "Zip" begin
```

Comment:
I'm slightly nervous that there's only one test here, given that we've got the dependence on `_unzip` which needs the `Zip` to be non-empty and for `length` to be defined over the things being iterated over.

It would be good to add some additional tests on objects without a defined length and empty `Zip`s, since we've got no way to exclude them via dispatch.